sqlflow/magic.py

```diff
@@ -42,9 +54,98 @@ def execute(self, line, cell):
+        self.create_db_on_demaond()
+        if not self.client:
+            self.client = Client()
+
+    def get_ssl_ctx(self):
+        ca_path = os.getenv("SQLFLOW_PLAYGROUND_SERVER_CA")
+        client_key = os.getenv("SQLFLOW_PLAYGROUND_CLIENT_KEY")
+        client_cert = os.getenv("SQLFLOW_PLAYGROUND_CLIENT_CERT")
+        if not (ca_path and client_cert and client_key):
+            raise ValueError("Certification files is not configured"
+                             "for this client")
+        ssl_ctx = ssl.create_default_context(ssl.Purpose.SERVER_AUTH)
+        ssl_ctx.load_cert_chain(client_cert, client_key)
+        ssl_ctx.load_verify_locations(ca_path)
+        ssl_ctx.check_hostname = False
+
+        return ssl_ctx
+
+    def create_db_on_demaond(self):
+        """If we are connecting to a Sqlflow Playground Service,
+        we need to ask the server to get DB and SQLFlow resource.
+        In this case, we do not specify SQLFLOW_DATASOURCE, instead
+        we should specify the SQLFLOW_PLAYGROUND_SERVRE
+        variable. On first time %%sqlflow command is executed,
+        we check if the DB connection is retrived from the server,
+        and create a new one if we haven't done it.
+
+        Client should be secured by a certification file. Use
+        SQLFLOW_PLAYGROUND_CLIENT_KEY and SQLFLOW_PLAYGROUND_CLIENT_CERT
+        and SQLFLOW_PLAYGROUND_SERVER_CA to specify them. These
+        files can be got from SQLFlow Playground Server maintainer.
+        """
+
+        if os.getenv("SQLFLOW_DATASOURCE"):
+            return
+        # If no datasource is given, try to connect to
+        # our playground server to create one
+        user_id_env = os.getenv("SQLFLOW_PLAYGROUND_USER_ID_ENV")
+        user_id = os.getenv(user_id_env)
+        server = os.getenv("SQLFLOW_PLAYGROUND_SERVRE")
+        if not server:
+            raise ValueError("Neither a datasource nor a "
+                             "playground server is given.")
+        if not user_id:
+            raise ValueError(
+                "Need to specify a SQLFLOW_PLAYGROUND_USER_ID_ENV")
+        # give user some hint, this may take a few seconds
+        from IPython.core.display import display, HTML
+        display(HTML("Loading resource..."))
+        # create db pod on playground service
+        body = {
+            "user_id": user_id,
+        }
+        ssl_ctx = self.get_ssl_ctx()
+        http = httpclient.HTTPClient()
+        try:
+            req = httpclient.HTTPRequest(
+                "%s/api/create_db" % server, method="POST",
+                ssl_options=ssl_ctx, body=json.dumps(body))
+            resp = http.fetch(req)
+            result = json.loads(resp.body)
+            os.environ["SQLFLOW_DATASOURCE"] = result["data_source"]
+
+            # server can kill the db resource
+            # if no heart beat is coming for a while
+            self.setup_heart_beat(server, user_id)
+        finally:
+            http.close()
```

Comment:
Do we need to raise an exception here to ensure the error message is returned to the notebook for display?

Reply:
done